In [1]:
# Install SQLAlchemy 1.4.4
!pip install sqlalchemy==1.4.4

# Install ipython-sql 0.4.1
!pip install ipython-sql==0.4.1

In [2]:
%load_ext sql

In [3]:
!pip install pymysql

In [4]:
%sql mysql+pymysql://root:DSEfe0507727182@127.0.0.1:3306/beers

In [6]:
%%sql

SELECT *
FROM beers
LIMIT 5

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
5 rows affected.


[(0, 0.05, None, 1436, 'Pub Beer', 'American Pale Lager', 408, 12.0),
 (1, 0.066, None, 2265, "Devil's Cup", 'American Pale Ale (APA)', 177, 12.0),
 (2, 0.071, None, 2264, 'Rise of the Phoenix', 'American IPA', 177, 12.0),
 (3, 0.09, None, 2263, 'Sinister', 'American Double / Imperial IPA', 177, 12.0),
 (4, 0.075, None, 2262, 'Sex and Candy', 'American IPA', 177, 12.0)]

In [7]:
%%sql

SELECT *
FROM breweries
LIMIT 5

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
5 rows affected.


[(0, 'NorthGate Brewing ', 'Minneapolis', ' MN'),
 (1, 'Against the Grain Brewery', 'Louisville', ' KY'),
 (2, "Jack's Abby Craft Lagers", 'Framingham', ' MA'),
 (3, 'Mike Hess Brewing Company', 'San Diego', ' CA'),
 (4, 'Fort Point Beer Company', 'San Francisco', ' CA')]

It seems the beer dataset contains data about the beer and the Breweries contains data about the breweries producing them.

the "abv" column is the alcoholic content by volume with 0 being no alcohol and 1 being pure alcohol, the "ibu" column is known as International bittering units, which describe how bitter a drink is. 
The "id" column is the unique id of each beer and the name column contains the name of the beer. The style has to do with the Beer style (lager, ale, IPA, etc.).
 the "brewery_id" column is the Unique identifier for the brewery that produces this beer; and can use to join with brewery info, finally, the "ounce" column is the Size of beer in ounces.

## the most alcoholic beer in the data set 

In [13]:
%%sql

SELECT *
FROM beers
ORDER BY abv DESC
LIMIT 10

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
10 rows affected.


[(2278, 0.128, '', 2565, 'Lee Hill Series Vol. 5 - Belgian Style Quadrupel Ale', 'Quadrupel (Quad)', 51, 19.2),
 (70, 0.125, '80.0', 2685, 'London Balling', 'English Barleywine', 1, 16.0),
 (2189, 0.12, '90.0', 2621, 'Csar', 'Russian Imperial Stout', 17, 16.0),
 (2279, 0.10400000000000001, '', 2564, 'Lee Hill Series Vol. 4 - Manhattan Style Rye Ale', 'Rye Beer', 51, 19.2),
 (1857, 0.1, '52.0', 2574, '4Beans', 'Baltic Porter', 46, 12.0),
 (1771, 0.099, '', 2249, 'Sanctified', 'Belgian Strong Pale Ale', 164, 12.0),
 (437, 0.099, '', 1033, 'Hubris Quadrupel Anniversary Ale', 'Quadrupel (Quad)', 10, 16.0),
 (2286, 0.099, '90.0', 1315, 'Upslope Imperial India Pale Ale', 'American Double / Imperial IPA', 51, 19.2),
 (1045, 0.099, '85.0', 904, 'Ex Umbris Rye Imperial Stout', 'American Double / Imperial Stout', 423, 16.0),
 (1243, 0.099, '', 2454, 'Galaxy High', 'American Double / Imperial IPA', 91, 12.0)]

The beer with the beer name Lee Hill Series and Style Quadrupel Ale	(abv = 0.128) is the most alcoholic beer in this dataset.

In [14]:
%%sql

SELECT *
FROM beers
ORDER BY abv
LIMIT 10

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
10 rows affected.


[(2266, 0.001, '', 606, 'Scotty K NA', 'Low Alcohol Beer', 522, 16.0),
 (769, 0.027000000000000003, '', 1286, 'Bikini Beer', 'American IPA', 173, 12.0),
 (1074, 0.027000000000000003, '21.0', 2482, 'Totally Radler', 'Radler', 80, 16.0),
 (2394, 0.027999999999999997, '15.0', 2306, 'Summer Brew', 'American Pilsner', 109, 12.0),
 (2027, 0.032, '13.0', 2231, 'American Light', 'Light Lager', 129, 12.0),
 (138, 0.032, '27.0', 550, 'Rocket Girl', 'Kölsch', 528, 12.0),
 (1860, 0.032, '7.0', 2266, 'Rad', 'Fruit / Vegetable Beer', 46, 16.0),
 (1460, 0.034, '6.0', 1417, 'Weiss Trash Culture', 'Berliner Weissbier', 410, 12.0),
 (190, 0.035, '45.0', 1643, 'Banner American Ale', 'American Amber / Red Ale', 318, 12.0),
 (42, 0.035, '11.0', 2545, 'Passion Fruit Prussia', 'Berliner Weissbier', 60, 12.0)]

The beer with the beer name Scotty K NA	 and StyleLow Alcohol Beer	(abv = 0.001) is the least alcoholic beer in this dataset.

In [16]:
%%sql

SELECT trim(*)
FROM beers
ORDER BY ibu DESC
LIMIT 3

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
3 rows affected.


[(1788, 0.091, '99.0', 2063, 'Count Hopula (Vault Series)', 'American Double / Imperial IPA', 30, 16.0),
 (2084, 0.062, '99.0', 12, 'Furious', 'American IPA', 61, 16.0),
 (579, 0.095, '99.0', 1647, 'Terrace Hill Double IPA', 'American Double / Imperial IPA', 350, 16.0)]

The beer with the beer name Count Hopula (Vault Series) and American Double / Imperial IPA (ibu = 99.0) is the most bitter alcoholic beer in this dataset.

In [19]:
%%sql

SELECT trim(*)
FROM beers
ORDER BY ibu DESC

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
(pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '*)\nFROM beers\nORDER BY ibu DESC' at line 1")
[SQL: SELECT trim(*)
FROM beers
ORDER BY ibu DESC]
(Background on this error at: http://sqlalche.me/e/14/f405)


In [20]:
%%sql

SELECT trim ibu
FROM beers
ORDER BY ibu DESC
LIMIT 10

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
(pymysql.err.OperationalError) (1054, "Unknown column 'trim' in 'field list'")
[SQL: SELECT trim ibu
FROM beers
ORDER BY ibu DESC
LIMIT 10]
(Background on this error at: http://sqlalche.me/e/14/e3q8)


In [21]:
%%sql

SELECT trim(ibu)
FROM beers
ORDER BY ibu DESC
LIMIT 10

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
10 rows affected.


[('99.0',),
 ('99.0',),
 ('99.0',),
 ('99.0',),
 ('98.0',),
 ('98.0',),
 ('98.0',),
 ('98.0',),
 ('97.0',),
 ('95.0',)]

In [22]:
%%sql
DESCRIBE beers

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
8 rows affected.


[('SerialN', 'int', 'YES', '', None, ''),
 ('abv', 'double', 'YES', '', None, ''),
 ('ibu', 'text', 'YES', '', None, ''),
 ('id', 'int', 'YES', '', None, ''),
 ('beers_name', 'text', 'YES', '', None, ''),
 ('style', 'text', 'YES', '', None, ''),
 ('brewery_id', 'int', 'YES', 'MUL', None, ''),
 ('ounces', 'double', 'YES', '', None, '')]

In [25]:
%%sql
ALTER TABLE beers
MODIFY ibu DOUBLE;

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
(pymysql.err.DataError) (1265, "Data truncated for column 'ibu' at row 1")
[SQL: ALTER TABLE beers
MODIFY ibu DOUBLE;]
(Background on this error at: http://sqlalche.me/e/14/9h9h)


In [33]:
%%sql
SELECT ibu 
FROM beers 
WHERE CAST(ibu AS DOUBLE) IS NULL OR ibu = ''
LIMIT 5;

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
5 rows affected.


[(None,), (None,), (None,), (None,), (None,)]

In [29]:
%%sql

SELECT *
FROM beers
LIMIT 5

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
5 rows affected.


[(0, 0.05, None, 1436, 'Pub Beer', 'American Pale Lager', 408, 12.0),
 (1, 0.066, None, 2265, "Devil's Cup", 'American Pale Ale (APA)', 177, 12.0),
 (2, 0.071, None, 2264, 'Rise of the Phoenix', 'American IPA', 177, 12.0),
 (3, 0.09, None, 2263, 'Sinister', 'American Double / Imperial IPA', 177, 12.0),
 (4, 0.075, None, 2262, 'Sex and Candy', 'American IPA', 177, 12.0)]

In [30]:
%%sql

SELECT trim(ibu)
FROM beers
ORDER BY ibu DESC
LIMIT 10

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
10 rows affected.


[('99.0',),
 ('99.0',),
 ('99.0',),
 ('99.0',),
 ('98.0',),
 ('98.0',),
 ('98.0',),
 ('98.0',),
 ('97.0',),
 ('95.0',)]

In [31]:
%%sql
ALTER TABLE beers
MODIFY ibu DOUBLE;

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
2348 rows affected.


ResourceClosedError: This result object does not return rows. It has been closed automatically.

In [32]:
%%sql

SELECT *
FROM beers
ORDER BY ibu DESC
LIMIT 10

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
10 rows affected.


[(147, 0.08199999999999999, 138.0, 980, 'Bitter Bitch Imperial IPA', 'American Double / Imperial IPA', 374, 12.0),
 (2390, 0.059000000000000004, 135.0, 1676, 'Troopers Alley IPA', 'American IPA', 344, 12.0),
 (525, 0.09, 130.0, 2067, 'Dead-Eye DIPA', 'American Double / Imperial IPA', 230, 16.0),
 (594, 0.08900000000000001, 126.0, 2440, 'Bay of Bengal Double IPA (2014)', 'American Double / Imperial IPA', 99, 12.0),
 (2144, 0.08, 120.0, 1111, 'Heady Topper', 'American Double / Imperial IPA', 272, 16.0),
 (2145, 0.08, 120.0, 379, 'Heady Topper', 'American Double / Imperial IPA', 272, 16.0),
 (2081, 0.09699999999999999, 120.0, 15, 'Abrasive Ale', 'American Double / Imperial IPA', 61, 16.0),
 (474, 0.09, 118.0, 2123, 'More Cowbell', 'American Double / Imperial IPA', 214, 16.0),
 (150, 0.075, 115.0, 2503, 'Hop A-Peel', 'American Double / Imperial IPA', 72, 16.0),
 (2191, 0.085, 115.0, 1449, 'Overlord Imperial IPA', 'American Double / Imperial IPA', 17, 16.0)]

The beer with the beer name Bitter Bitch Imperial IP and American Double / Imperial IPA (ibu = 138.0) is the most bitter alcoholic beer in this dataset.

In [10]:
%%sql
SELECT *
FROM breweries
LIMIT 10

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
10 rows affected.


[(0, 'NorthGate Brewing ', 'Minneapolis', ' MN'),
 (1, 'Against the Grain Brewery', 'Louisville', ' KY'),
 (2, "Jack's Abby Craft Lagers", 'Framingham', ' MA'),
 (3, 'Mike Hess Brewing Company', 'San Diego', ' CA'),
 (4, 'Fort Point Beer Company', 'San Francisco', ' CA'),
 (5, 'COAST Brewing Company', 'Charleston', ' SC'),
 (6, 'Great Divide Brewing Company', 'Denver', ' CO'),
 (7, 'Tapistry Brewing', 'Bridgman', ' MI'),
 (8, 'Big Lake Brewing', 'Holland', ' MI'),
 (9, 'The Mitten Brewing Company', 'Grand Rapids', ' MI')]

In [12]:
%%sql

SELECT beers.ibu, beers.brewery_id, beers.beers_name, breweries.brewery_id, breweries.brewery_name
FROM beers
INNER JOIN breweries
ON beers.brewery_id = breweries.brewery_id
ORDER BY beers.ibu DESC
LIMIT 5

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
5 rows affected.


[(138.0, 374, 'Bitter Bitch Imperial IPA', 374, 'Astoria Brewing Company'),
 (135.0, 344, 'Troopers Alley IPA', 344, 'Wolf Hills Brewing Company'),
 (130.0, 230, 'Dead-Eye DIPA', 230, 'Cape Ann Brewing Company'),
 (126.0, 99, 'Bay of Bengal Double IPA (2014)', 99, 'Christian Moerlein Brewing Company'),
 (120.0, 61, 'Abrasive Ale', 61, 'Surly Brewing Company')]

The beer with the beer name Bitter Bitch Imperial IPA and and produced by the Astoria Brewing Company (ibu = 138.0) is the most bitter alcoholic beer in this combined data sets.

In [13]:
%%sql

SELECT be.ibu, be.brewery_id, be.beers_name, br.brewery_id, br.brewery_name
FROM beers AS be
INNER JOIN breweries AS br
ON be.brewery_id = br.brewery_id
ORDER BY be.ibu DESC
LIMIT 5

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
5 rows affected.


[(138.0, 374, 'Bitter Bitch Imperial IPA', 374, 'Astoria Brewing Company'),
 (135.0, 344, 'Troopers Alley IPA', 344, 'Wolf Hills Brewing Company'),
 (130.0, 230, 'Dead-Eye DIPA', 230, 'Cape Ann Brewing Company'),
 (126.0, 99, 'Bay of Bengal Double IPA (2014)', 99, 'Christian Moerlein Brewing Company'),
 (120.0, 61, 'Abrasive Ale', 61, 'Surly Brewing Company')]

The above was repeated in the alliance format with be representing beers and br rep breweries. Hence, they were replaced in the synthax

In [3]:
%%sql

SELECT DISTINCT (city)
FROM breweries
LIMIT 10

Environment variable $DATABASE_URL not set, and no connect string given.
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys([])


In [30]:
%%sql

SELECT DISTINCT (state)
FROM breweries
LIMIT 10

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
10 rows affected.


[(' MN',),
 (' KY',),
 (' MA',),
 (' CA',),
 (' SC',),
 (' CO',),
 (' MI',),
 (' IN',),
 (' TX',),
 (' AZ',)]

In [16]:
%%sql

SELECT *
FROM beers
LIMIT 2

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
2 rows affected.


[(0, 0.05, None, 1436, 'Pub Beer', 'American Pale Lager', 408, 12.0),
 (1, 0.066, None, 2265, "Devil's Cup", 'American Pale Ale (APA)', 177, 12.0)]

In [4]:
%%sql

SELECT be.brewery_id, br.brewery_id, br.brewery_name, be.beers_name, be.style, br.state
FROM beers AS be
INNER JOIN breweries AS br
ON be.brewery_id = br.brewery_id
WHERE state LIKE '%CA%';

Environment variable $DATABASE_URL not set, and no connect string given.
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys([])


In [31]:
%%sql

SELECT br.brewery_id, br.brewery_name, be.beers_name, be.style, br.state
FROM beers AS be
INNER JOIN breweries AS br
ON be.brewery_id = br.brewery_id
WHERE state LIKE "%CA%";

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
182 rows affected.


[(3, 'Mike Hess Brewing Company', 'Habitus (2014)', 'American Double / Imperial IPA', ' CA'),
 (3, 'Mike Hess Brewing Company', 'Solis', 'American IPA', ' CA'),
 (3, 'Mike Hess Brewing Company', 'Jucundus', 'Wheat Ale', ' CA'),
 (3, 'Mike Hess Brewing Company', 'Habitus', 'American Double / Imperial IPA', ' CA'),
 (3, 'Mike Hess Brewing Company', 'Grazias', 'Cream Ale', ' CA'),
 (3, 'Mike Hess Brewing Company', 'Claritas', 'Kölsch', ' CA'),
 (4, 'Fort Point Beer Company', 'Park', 'American Pale Wheat Ale', ' CA'),
 (4, 'Fort Point Beer Company', 'Westfalia', 'American Amber / Red Ale', ' CA'),
 (4, 'Fort Point Beer Company', 'KSA', 'Kölsch', ' CA'),
 (4, 'Fort Point Beer Company', 'Villager', 'American IPA', ' CA'),
 (33, "The Dudes' Brewing Company", 'Surfrider', 'American Pale Ale (APA)', ' CA'),
 (33, "The Dudes' Brewing Company", 'Kolschtal Eddy', 'Kölsch', ' CA'),
 (33, "The Dudes' Brewing Company", 'South Bay Session IPA', 'American IPA', ' CA'),
 (33, "The Dudes' Brewing Company", "Grandma's Pecan", 'English Brown Ale', ' CA'),
 (33, "The Dudes' Brewing Company", 'Double Trunk', 'American Double / Imperial IPA', ' CA'),
 (34, 'Ballast Point Brewing Company', 'Grapefruit Sculpin', 'American IPA', ' CA'),
 (34, 'Ballast Point Brewing Company', 'Even Keel', 'American IPA', ' CA'),
 (34, 'Ballast Point Brewing Company', 'Ballast Point Pale Ale', 'Kölsch', ' CA'),
 (34, 'Ballast Point Brewing Company', 'Big Eye India Pale Ale', 'American IPA', ' CA'),
 (34, 'Ballast Point Brewing Company', 'Longfin Lager', 'Munich Helles Lager', ' CA'),
 (34, 'Ballast Point Brewing Company', 'Sculpin IPA', 'American IPA', ' CA'),
 (35, 'Anchor Brewing Company', 'Liberty Ale', 'American IPA', ' CA'),
 (35, 'Anchor Brewing Company', 'IPA', 'American IPA', ' CA'),
 (35, 'Anchor Brewing Company', 'Summer Wheat', 'American Pale Wheat Ale', ' CA'),
 (35, 'Anchor Brewing Company', 'California Lager', 'American Amber / Red Lager', ' CA'),
 (35, 'Anchor Brewing Company', 'Brotherhood Steam', 'California Common / Steam Beer', ' CA'),
 (36, 'Figueroa Mountain Brewing Company', 'Weiss Weiss Baby', 'Kristalweizen', ' CA'),
 (36, 'Figueroa Mountain Brewing Company', 'Czech Yo Self', 'Czech Pilsener', ' CA'),
 (36, 'Figueroa Mountain Brewing Company', 'FMB 101', 'Kölsch', ' CA'),
 (48, 'Firestone Walker Brewing Company', 'Easy Jack', 'American IPA', ' CA'),
 (48, 'Firestone Walker Brewing Company', 'Union Jack', 'American IPA', ' CA'),
 (48, 'Firestone Walker Brewing Company', 'Pivo Pils', 'German Pilsener', ' CA'),
 (48, 'Firestone Walker Brewing Company', '805 Blonde Ale', 'American Blonde Ale', ' CA'),
 (48, 'Firestone Walker Brewing Company', '805', 'American Blonde Ale', ' CA'),
 (78, 'Speakasy Ales & Lagers', 'Baby Daddy Session IPA', 'American IPA', ' CA'),
 (83, 'Sierra Nevada Brewing Company', 'Nooner', 'German Pilsener', ' CA'),
 (83, 'Sierra Nevada Brewing Company', 'Torpedo', 'American IPA', ' CA'),
 (83, 'Sierra Nevada Brewing Company', 'Yonder Bock', 'Maibock / Helles Bock', ' CA'),
 (83, 'Sierra Nevada Brewing Company', 'CANfusion Rye Bock', 'Rye Beer', ' CA'),
 (83, 'Sierra Nevada Brewing Company', 'Sierra Nevada Pale Ale', 'American Pale Ale (APA)', ' CA'),
 (83, 'Sierra Nevada Brewing Company', 'Old Chico Crystal Wheat', 'American Pale Wheat Ale', ' CA'),
 (83, 'Sierra Nevada Brewing Company', 'Summerfest', 'Czech Pilsener', ' CA'),
 (83, 'Sierra Nevada Brewing Company', 'Torpedo', 'American IPA', ' CA'),
 (83, 'Sierra Nevada Brewing Company', 'Sierra Nevada Pale Ale', 'American Pale Ale (APA)', ' CA'),
 (98, 'Mission Brewery', 'Cortez Gold', 'Belgian Pale Ale', ' CA'),
 (98, 'Mission Brewery', 'Mission IPA', 'American IPA', ' CA'),
 (98, 'Mission Brewery', 'El Conquistador Extra Pale Ale', 'American Pale Ale (APA)', ' CA'),
 (98, 'Mission Brewery', 'Shipwrecked Double IPA', 'American Double / Imperial IPA', ' CA'),
 (111, 'Mother Earth Brew Company', "Cali Creamin'", 'Cream Ale', ' CA'),
 (112, 'Black Market Brewing Company', 'Decep

NOTE: There are 182 beer

In [22]:
%%sql

SELECT DISTINCT be.beers_name, br.brewery_id, br.brewery_name, be.style, br.state
FROM beers AS be
INNER JOIN breweries AS br
ON be.brewery_id = br.brewery_id
WHERE state LIKE "%CA%";


 * mysql+pymysql://root:***@127.0.0.1:3306/beers
173 rows affected.
(pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'LIMIT 5' at line 1")
[SQL: LIMIT 5]
(Background on this error at: http://sqlalche.me/e/14/f405)


The distribution of the various beer names in CA in the join data of breweries and beers. 
Table appearing in the order of placement/arrangement of what were selected.
The DISTINCT FUNCTION ANLY AFFECTED THE BEER NAME. Thus, reducing the number of rows from 182 to 173.

In [36]:
%%sql

SELECT DISTINCT br.brewery_name, be.beers_name, br.brewery_id, be.style, br.state
FROM beers AS be
INNER JOIN breweries AS br
ON be.brewery_id = br.brewery_id
WHERE state LIKE "%CA%";

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
173 rows affected.


[('Mike Hess Brewing Company', 'Habitus (2014)', 3, 'American Double / Imperial IPA', ' CA'),
 ('Mike Hess Brewing Company', 'Solis', 3, 'American IPA', ' CA'),
 ('Mike Hess Brewing Company', 'Jucundus', 3, 'Wheat Ale', ' CA'),
 ('Mike Hess Brewing Company', 'Habitus', 3, 'American Double / Imperial IPA', ' CA'),
 ('Mike Hess Brewing Company', 'Grazias', 3, 'Cream Ale', ' CA'),
 ('Mike Hess Brewing Company', 'Claritas', 3, 'Kölsch', ' CA'),
 ('Fort Point Beer Company', 'Park', 4, 'American Pale Wheat Ale', ' CA'),
 ('Fort Point Beer Company', 'Westfalia', 4, 'American Amber / Red Ale', ' CA'),
 ('Fort Point Beer Company', 'KSA', 4, 'Kölsch', ' CA'),
 ('Fort Point Beer Company', 'Villager', 4, 'American IPA', ' CA'),
 ("The Dudes' Brewing Company", 'Surfrider', 33, 'American Pale Ale (APA)', ' CA'),
 ("The Dudes' Brewing Company", 'Kolschtal Eddy', 33, 'Kölsch', ' CA'),
 ("The Dudes' Brewing Company", 'South Bay Session IPA', 33, 'American IPA', ' CA'),
 ("The Dudes' Brewing Company", "Grandma's Pecan", 33, 'English Brown Ale', ' CA'),
 ("The Dudes' Brewing Company", 'Double Trunk', 33, 'American Double / Imperial IPA', ' CA'),
 ('Ballast Point Brewing Company', 'Grapefruit Sculpin', 34, 'American IPA', ' CA'),
 ('Ballast Point Brewing Company', 'Even Keel', 34, 'American IPA', ' CA'),
 ('Ballast Point Brewing Company', 'Ballast Point Pale Ale', 34, 'Kölsch', ' CA'),
 ('Ballast Point Brewing Company', 'Big Eye India Pale Ale', 34, 'American IPA', ' CA'),
 ('Ballast Point Brewing Company', 'Longfin Lager', 34, 'Munich Helles Lager', ' CA'),
 ('Ballast Point Brewing Company', 'Sculpin IPA', 34, 'American IPA', ' CA'),
 ('Anchor Brewing Company', 'Liberty Ale', 35, 'American IPA', ' CA'),
 ('Anchor Brewing Company', 'IPA', 35, 'American IPA', ' CA'),
 ('Anchor Brewing Company', 'Summer Wheat', 35, 'American Pale Wheat Ale', ' CA'),
 ('Anchor Brewing Company', 'California Lager', 35, 'American Amber / Red Lager', ' CA'),
 ('Anchor Brewing Company', 'Brotherhood Steam', 35, 'California Common / Steam Beer', ' CA'),
 ('Figueroa Mountain Brewing Company', 'Weiss Weiss Baby', 36, 'Kristalweizen', ' CA'),
 ('Figueroa Mountain Brewing Company', 'Czech Yo Self', 36, 'Czech Pilsener', ' CA'),
 ('Figueroa Mountain Brewing Company', 'FMB 101', 36, 'Kölsch', ' CA'),
 ('Firestone Walker Brewing Company', 'Easy Jack', 48, 'American IPA', ' CA'),
 ('Firestone Walker Brewing Company', 'Union Jack', 48, 'American IPA', ' CA'),
 ('Firestone Walker Brewing Company', 'Pivo Pils', 48, 'German Pilsener', ' CA'),
 ('Firestone Walker Brewing Company', '805 Blonde Ale', 48, 'American Blonde Ale', ' CA'),
 ('Firestone Walker Brewing Company', '805', 48, 'American Blonde Ale', ' CA'),
 ('Speakasy Ales & Lagers', 'Baby Daddy Session IPA', 78, 'American IPA', ' CA'),
 ('Sierra Nevada Brewing Company', 'Nooner', 83, 'German Pilsener', ' CA'),
 ('Sierra Nevada Brewing Company', 'Torpedo', 83, 'American IPA', ' CA'),
 ('Sierra Nevada Brewing Company', 'Yonder Bock', 83, 'Maibock / Helles Bock', ' CA'),
 ('Sierra Nevada Brewing Company', 'CANfusion Rye Bock', 83, 'Rye Beer', ' CA'),
 ('Sierra Nevada Brewing Company', 'Sierra Nevada Pale Ale', 83, 'American Pale Ale (APA)', ' CA'),
 ('Sierra Nevada Brewing Company', 'Old Chico Crystal Wheat', 83, 'American Pale Wheat Ale', ' CA'),
 ('Sierra Nevada Brewing Company', 'Summerfest', 83, 'Czech Pilsener', ' CA'),
 ('Mission Brewery', 'Cortez Gold', 98, 'Belgian Pale Ale', ' CA'),
 ('Mission Brewery', 'Mission IPA', 98, 'American IPA', ' CA'),
 ('Mission Brewery', 'El Conquistador Extra Pale Ale', 98, 'American Pale Ale (APA)', ' CA'),
 ('Mission Brewery', 'Shipwrecked Double IPA', 98, 'American Double / Imperial IPA', ' CA'),
 ('Mother Earth Brew Company', "Cali Creamin'", 111, 'Cream Ale', ' CA'),
 ('Black Market Brewing Company', 'Deception', 112, 'American Blonde Ale', ' CA'),
 ('Black Market Brewing Company', 'Blackmarket Rye IPA', 112, 'American IPA', ' CA'),
 ('Black Market Brewing Company', 'Black Market He

The distribution of the various beer names in CA in the join data of breweries and beers. 
Table appearing in the order of placement/arrangement of what were selected.
The DISTINCT FUNCTION ANLY AFFECTED THE BEER NAME. Thus, reducing the number of rows from 182 to 173.

In [49]:
%%sql

SELECT *
FROM breweries
LIMIT 10

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
10 rows affected.


[(0, 'NorthGate Brewing ', 'Minneapolis', ' MN'),
 (1, 'Against the Grain Brewery', 'Louisville', ' KY'),
 (2, "Jack's Abby Craft Lagers", 'Framingham', ' MA'),
 (3, 'Mike Hess Brewing Company', 'San Diego', ' CA'),
 (4, 'Fort Point Beer Company', 'San Francisco', ' CA'),
 (5, 'COAST Brewing Company', 'Charleston', ' SC'),
 (6, 'Great Divide Brewing Company', 'Denver', ' CO'),
 (7, 'Tapistry Brewing', 'Bridgman', ' MI'),
 (8, 'Big Lake Brewing', 'Holland', ' MI'),
 (9, 'The Mitten Brewing Company', 'Grand Rapids', ' MI')]

In [21]:
%%sql

SELECT state, city, brewery_name
FROM breweries
ORDER BY brewery_name DESC
LIMIT 3

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
3 rows affected.


[(' CO', 'Denver', 'Wynkoop Brewing Company'),
 (' OR', 'Bend', 'Worthy Brewing Company'),
 (' MA', 'Worcester', 'Wormtown Brewery')]

In [5]:
%%sql

SELECT city, state, COUNT(city) as num_of_city
FROM breweries
GROUP BY city, state
ORDER BY num_of_city DESC
LIMIT 10

Environment variable $DATABASE_URL not set, and no connect string given.
Connection info needed in SQLAlchemy format, example:
               postgresql://username:password@hostname/dbname
               or an existing connection: dict_keys([])


In [20]:
%%sql

SELECT state, city, COUNT(city) as num_of_city
FROM breweries
GROUP BY city, state
ORDER BY num_of_city DESC
LIMIT 3

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
3 rows affected.


[(' OR', 'Portland', 11), (' CO', 'Boulder', 9), (' IL', 'Chicago', 9)]

The above command helps to determine the frequency distribution of the Cities in the different State from the breweries data set.

%%sql

SELECT state, city, COUNT(city) as num_of_city
FROM breweries
GROUP BY city, state
ORDER BY num_of_city DESC

In [ ]:
%%sql

SELECT state, city, COUNT(city) as num_of_city
FROM breweries
GROUP BY city, state
ORDER BY num_of_city DESC

In [6]:
%%sql 

SELECT *
FROM beers
LIMIT 4

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
4 rows affected.


[(0, 0.05, None, 1436, 'Pub Beer', 'American Pale Lager', 408, 12.0),
 (1, 0.066, None, 2265, "Devil's Cup", 'American Pale Ale (APA)', 177, 12.0),
 (2, 0.071, None, 2264, 'Rise of the Phoenix', 'American IPA', 177, 12.0),
 (3, 0.09, None, 2263, 'Sinister', 'American Double / Imperial IPA', 177, 12.0)]

In [19]:
%%sql

SELECT brewery_id, beers_name, COUNT(beers_name) as num_of_beers_name
FROM beers
GROUP BY beers_name, brewery_id
ORDER BY num_of_beers_name DESC
LIMIT 3

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
3 rows affected.


[(80, 'Nonstop Hef Hop', 12),
 (439, 'Longboard Island Lager', 4),
 (166, "Dale's Pale Ale", 4)]

In [18]:
%%sql

SELECT beers.brewery_id, breweries.state, breweries.brewery_name, beers.beers_name, COUNT(beers.beers_name) as num_of_beers_name
FROM beers
INNER JOIN breweries
ON beers.brewery_id = breweries.brewery_id
GROUP BY beers.brewery_id, breweries.state, breweries.brewery_name, beers.beers_name
ORDER BY num_of_beers_name DESC
LIMIT 3

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
3 rows affected.


[(80, ' OR', 'Hopworks Urban Brewery', 'Nonstop Hef Hop', 12),
 (166, ' CO', 'Oskar Blues Brewery', "Dale's Pale Ale", 4),
 (439, ' HI', 'Kona Brewing Company', 'Longboard Island Lager', 4)]

Note: The combined data revealed that the beers (name) with the highest occurence is Nonstop Hef Hop with a frequency of 12 and it is produced by Hopworks Urban Brewery	

In [17]:
%%sql

SELECT state, brewery_name, COUNT(brewery_name) as freq_of_brewery
FROM breweries
GROUP BY state, brewery_name
ORDER BY freq_of_brewery DESC
LIMIT 4

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
4 rows affected.


[(' VT', 'Otter Creek Brewing', 2),
 (' KY', 'Against the Grain Brewery', 2),
 (' MN', 'Summit Brewing Company', 2),
 (' PA', 'Sly Fox Brewing Company', 2)]

In [15]:
%%sql

SELECT *
FROM breweries
LIMIT 3

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
3 rows affected.


[(0, 'NorthGate Brewing ', 'Minneapolis', ' MN'),
 (1, 'Against the Grain Brewery', 'Louisville', ' KY'),
 (2, "Jack's Abby Craft Lagers", 'Framingham', ' MA')]

In [23]:
%%sql

SELECT state, brewery_name, COUNT(state) as freq_of_state
FROM breweries
GROUP BY state, brewery_name
ORDER BY freq_of_state DESC
LIMIT 4

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
4 rows affected.


[(' VT', 'Otter Creek Brewing', 2),
 (' KY', 'Against the Grain Brewery', 2),
 (' MN', 'Summit Brewing Company', 2),
 (' PA', 'Sly Fox Brewing Company', 2)]

In [27]:
%%sql

SELECT state, COUNT(state) as freq_of_state
FROM breweries
GROUP BY state
ORDER BY freq_of_state DESC
LIMIT 4

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
4 rows affected.


[(' CO', 47), (' CA', 39), (' MI', 32), (' OR', 29)]

In [32]:
%%sql

SELECT breweries.brewery_id, breweries.brewery_name, COUNT(beers.beers_name) as freq_of_beers_name
FROM beers
INNER JOIN breweries
ON beers.brewery_id = breweries.brewery_id
GROUP BY breweries.brewery_name, breweries.brewery_id
ORDER BY freq_of_beers_name DESC
LIMIT 3

 * mysql+pymysql://root:***@127.0.0.1:3306/beers
3 rows affected.


[(10, 'Brewery Vivant', 60),
 (25, 'Sun King Brewing Company', 36),
 (166, 'Oskar Blues Brewery', 29)]

Note: Above is the Brewery with the highest frequency of beers in the joined data set.